docs/posts/pydata-performance/index.qmd

```diff
@@ -0,0 +1,696 @@
+    [
+        df.uploaded_on.dt.floor("M"),
+        df.path.str.extract(r"\.([a-z0-9]+)$", 0),
+    ]
+)
+df
+```
+
+Here we hit the first API issue going back to an [old pandas
+issue](https://github.com/pandas-dev/pandas/issues/15303): we can't truncate
+a timestamp column to month frequency.
+
+Let's try the solution recommended in that issue.
+
+```{python}
+#| error: false
+%%time
+df = df.groupby(
+    [
+        df.uploaded_on.dt.to_period("M").dt.to_timestamp(),
+        df.path.str.extract(r"\.([a-z0-9]+)$", 0)
+    ]
+)
+df
+```
+
+That seems to have gotten past the `floor` issue but now I get an error about
+the shape of the grouping keys. I suspect this is because `extract` returns
+a `DataFrame`. Such is life.
+
+Let's try a different strategy: compute the columns in an `assign` call and
+then do the group by using string column names:
+
+```{python}
+%%time
+df = (
+    df.assign(
+        month=lambda t: t.uploaded_on.dt.to_period("M").dt.to_timestamp(), # <1>
+        ext=lambda t: t.path.str.extract(r"\.([a-z0-9]+)$", 0), # <1>
+    )
+    .groupby(["month", "ext"])
+    .agg({"project_name": lambda s: list(set(s))})
+    .rename(columns={"project_name": "projects"})
+)
+df
+```
+
+1. Later we'll see that Dask doesn't support lambda functions in `assign` and
+   instead we'll create an intermediate `DataFrame` for the `assign` call.
+
+Sort the values, add a new column and do the final aggregation:
+
+```{python}
+%%time
+df = (
+    df.reset_index()  # <1>
+    .sort_values("month", ascending=False)
+    .assign(
+        ext=lambda t: t.ext.str.replace(r"cxx|cpp|cc|c|hpp|h", "C/C++", regex=True)
+        .str.replace("^f.*$", "Fortran", regex=True)
+        .str.replace("rs", "Rust")
+        .str.replace("go", "Go")
+        .str.replace("asm", "Assembly")
+        .replace("", None)
+    )
+    .groupby(["month", "ext"])
+    .agg({"projects": lambda s: len(set(sum(s, [])))})
+)
+df
+```
+
+1. This call to `reset_index()` is necessary to avoid jumping through hoops to
+   do computation on a `MultiIndex` level. This is one of the trade-offs
+   associated with exposing indexes to the user--here as the keys of
+   a grouped aggregation: **grouping keys require a completely different API if
+   used in subsequent computations**.
+
+```{python}
+#| echo: false
+#
+# clean up after ourselves or else Dask will cause the kernel die from
+# running out of memory
+import gc
+
+del df
+gc.collect();  # semicolon to avoid printing the output
+```
+
+Remember, all of the previous code is executing on **a single file** and is
+taking close to three minutes to run.
+
+#### Conclusion
+
+If I only have pandas at my disposal, I'm unsure of how I can avoid getting
+a bigger computer to run this query over the entire data set.
+
+### Rewriting the query to be fair
+
+At this point I wondered whether this was a fair query to compare Dask and
+Pandas against.
+
+After all, Pandas' (and therefore Dask's) use of object arrays and their
```

Comment:
That's not correct anymore, Dask will use arrow backed strings by default for string columns since 2023.7.1

Reply:
The comment here is correct: Pandas and Dask's use of object arrays is inefficient, because object arrays are inefficient.

This comment is not about what the default is or should be or could be

Reply:
Are we talking about strings or general nested data structure? That's not totally clear to me.

It is correct for general nested data structure, but Dask doesn't use object arrays anymore for strings except if you explicitly ask for it

Reply:
Happy to add something like "pyarrow strings have been the default in dask since 2023.7.1" if that helps

Reply:
This is actually referring to list arrays, so I will adjust the language accordingly and I won't mention string arrays since they're not particularly relevant for the section.